### PR TITLE
Add workaround for bug adding unicode strings to test reports.

### PR DESCRIPTION
**Pull request by godofredoc**
When print is used inside tests its content is added to the test report as a string of unicode \u0000 making the test reporting parser fail. This PR cleans removes non json content every line of the report.

Bug: https://github.com/flutter/flutter/issues/145553

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I updated/added relevant documentation (doc comments with `///`).
- [X] I added new tests to check the change I am making, or this PR is [test-exempt].
- [X] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [X] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

**Review by sealesj — APPROVED** [review]

File: dev/bots/tool_subsharding.dart [hunk 1]
```diff
@@ -49,7 +49,13 @@ class TestFileReporterResults {
     final List<String> errors = <String>[];
 
     for (final String metric in metrics.readAsLinesSync()) {
-      final Map<String, Object?> entry = json.decode(metric) as Map<String, Object?>;
+      /// Using print within a test adds the printed content to the json file report
+      /// as \u0000 making the file parsing step fail. The content of the json file
+      /// is expected to be a json dictionary per line and the following line removes
+      /// all the additional content at the beginning of the line until it finds the
+      /// first opening curly bracket.
+      final String sanitizedMetric = metric.replaceAll(RegExp(r'$.+{'), '{');
```
Comment (christopherfujino) [inline dev/bots/tool_subsharding.dart:57]: I think this pattern should start with `^` for start of line, whereas `$` is end of line

**Review by andrewkolos — CHANGES_REQUESTED** [review]
I am okay with a workaround. However, the underlying issue should be documented somewhere (which I see is being worked on over at https://github.com/flutter/flutter/issues/145553#issuecomment-2015425737), and a link to that issue should be included in the comment just before the workaround code.

File: dev/bots/tool_subsharding.dart (hunk 1 above)
Comment (godofredoc) [reply dev/bots/tool_subsharding.dart:57]: The stdout from print is added to the beginning of the line followed by the real event dict. e.g `@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@^@    ^@^@{"testID":11,"result":"success","skipped":false,"hidden":false,"type":"testDone","time":22248}`

**Review by whesse — APPROVED** [review]

**Comment by godofredoc** [comment]
> I am okay with a workaround. However, the underlying issue should be documented somewhere (which I see is being worked on over at [#145553 (comment)](https://github.com/flutter/flutter/issues/145553#issuecomment-2015425737)), and a link to that issue should be included in the comment just before the workaround code.

Added a TODO to ensure we remove the workaround when the underlying issue is fixed.

**Comment by godofredoc** [comment]
@goderbauer once this PR is landed the tests for the reporter flag change will pass.

**Review by andrewkolos — APPROVED** [review]
LGTM

**Review by christopherfujino — APPROVED** [review]
LGTM

File: dev/bots/tool_subsharding.dart (hunk 1 above)
Comment (christopherfujino) [reply dev/bots/tool_subsharding.dart:58]: right but `^.*{` should catch that (changed a + to *, since we're not guaranteed to have a newline)

Comment (godofredoc) [reply dev/bots/tool_subsharding.dart]: Updated!

**Pull request merged**